--- gcc-3.4.4.old/./gcc/config/sparc/sparc-protos.h	2020-02-28 17:30:18.581569308 +0100
+++ gcc-3.4.4.new/./gcc/config/sparc/sparc-protos.h	2010-02-08 14:08:49.821098000 +0100
@@ -89,6 +89,14 @@
 extern int sparc_absnegfloat_split_legitimate (rtx, rtx);
 extern const char *output_ubranch (rtx, int, rtx);
 extern char *output_cbranch (rtx, rtx, int, int, int, int, rtx);
+extern char *output_divdf3 (rtx , rtx , rtx , rtx );
+extern void output_divdf3_emit (rtx , rtx , rtx , rtx );
+extern char *output_divsf3 (rtx , rtx , rtx , rtx );
+extern void output_divsf3_emit (rtx , rtx , rtx , rtx );
+extern char *output_sqrtdf2 (rtx , rtx , rtx );
+extern char *output_sqrtsf2 (rtx , rtx , rtx );
+extern void output_sqrtdf2_emit (rtx , rtx , rtx );
+extern void output_sqrtsf2_emit (rtx , rtx , rtx );
 extern const char *output_sibcall (rtx, rtx);
 extern char *output_v9branch (rtx, rtx, int, int, int, int, int, rtx);
 extern void emit_v9_brxx_insn (enum rtx_code, rtx, rtx);
